build-and-deploy-docs/create_mkdocs_config.py

```diff
@@ -0,0 +1,342 @@
+                        path=str(resolved_path.relative_to(docs_dir))
+                    )
+
+                # If the link is to an image, copy that image to the docs
+                elif magic.from_file(resolved_path, mime=True) \
+                        in VALID_IMAGE_MIME_TYPES:
+                    output_path = Path(img_dir, resolved_path.name)
+                    shutil.copy2(resolved_path, output_path)
+
+                    link = link._replace(
+                        path=str(output_path.relative_to(docs_dir))
+                    )
+
+                else:
+                    # For everything else, link to the file on GitHub
+                    link = link._replace(
+                        scheme="https",
+                        netloc="github.com",
+                        path=str(Path(
+                            pipeline_repo,
+                            "blob",
+                            os.environ.get("GITHUB_SHA", "main"),
+                            resolved_path.relative_to(readme_file.parent)
+                        ))
+                    )
+
+        elif link.fragment:
+            # This is an anchor link. As we've split the monolithic README into
+            # multiple files, we need to prepend those filepaths.
+            link = link._replace(
+                path=anchor_pages[get_heading_anchor(link.fragment)],
+            )
+
+        return urlunparse(link)
+
+    # Iterate through all tokens (including child tokens) to rewrite links and
+    # copy images into the docs/ folder
+    tokens_to_examine = collections.deque(tokens)
+    while tokens_to_examine:
+        token = tokens_to_examine.popleft()
+        if token.children:
+            tokens_to_examine.extend(token.children)
+
+        if token.type == "link_open" and "href" in token.attrs:
+            token.attrs["href"] = sanitize_link(token.attrs["href"])
+
+        elif token.type == "image" and "src" in token.attrs:
+            token.attrs["src"] = sanitize_link(token.attrs["src"])
+
+    # Write out each page to a separate file
+    renderer = mdformat.renderer.MDRenderer()
+    options = {
+        'parser_extension': [
+            mdformat.plugins.PARSER_EXTENSIONS['gfm'],
+            mdformat.plugins.PARSER_EXTENSIONS['tables'],
+        ]
+    }
+
+    table_of_contents = []
+
+    for page in pages:
+        Path(docs_dir, page.get_filename()).write_text(
+            renderer.render(page.tokens, options, {}),
+            encoding="utf-8"
+        )
+
+        table_of_contents.append({page.title: page.get_filename()})
+
+    return table_of_contents
+
+
+def get_mkdocs_config_data(path: Path, repo: str):
+    """ Read the given MKDocs config file or create it from default.
+
+    Args:
+        - `path`: Path to the MKDocs config file. When set to None, default
+          config is used.
+        - `repo`: The github repo name.
+    """
+    pipeline_name = repo.rsplit("/", maxsplit=1)[-1]
```

Comment:
Does the code run correctly if `repo` is `/foo/bar/` (trailing `/`)? `-1` would be `''` here.

Reply:
No, it would fail in that case - this is assuming a well-crafted `GITHUB_REPOSITORY` environment variable. I'll harden it up a bit.